neqo-http3/src/push_controller.rs

```diff
@@ -25,15 +96,333 @@ impl Display for PushController {
+                }
+                PushState::PushPromise { headers } | PushState::Active { headers, .. } => {
+                    if header_block != *headers {
+                        return Err(Error::HttpGeneralProtocol);
+                    }
+                    self.conn_events
+                        .push_promise(push_id, ref_stream_id, header_block);
+                    Ok(())
+                }
+                PushState::OnlyPushStream { stream_id, events } => {
+                    let stream_id_tmp = *stream_id;
+                    self.conn_events
+                        .push_promise(push_id, ref_stream_id, header_block.clone());
+
+                    for e in events.drain(..) {
+                        self.conn_events.insert(e);
+                    }
+                    *push_state = PushState::Active {
+                        stream_id: stream_id_tmp,
+                        headers: header_block,
+                    };
+                    Ok(())
+                }
+            },
+        }
+    }
+
+    pub fn add_new_push_stream(&mut self, push_id: u64, stream_id: u64) -> Res<bool> {
+        qtrace!(
+            "A new push stream with push_id={} stream_id={}",
+            push_id,
+            stream_id
+        );
+
+        self.check_push_id_and_create_stream_state(push_id)?;
+
+        match self.push_streams.get_mut(&push_id) {
+            None => {
+                qinfo!("Push has been closed already.");
+                Ok(false)
+            }
+            Some(push_state) => match push_state {
+                PushState::Init => {
+                    *push_state = PushState::OnlyPushStream {
+                        stream_id,
+                        events: Vec::new(),
+                    };
+                    Ok(true)
+                }
+                PushState::PushPromise { headers } => {
+                    let tmp = mem::replace(headers, Vec::new());
+                    *push_state = PushState::Active {
+                        stream_id,
+                        headers: tmp,
+                    };
+                    Ok(true)
+                }
+                // The following state have already have a push stream:
+                // PushState::OnlyPushStream | PushState::Active
+                _ => {
+                    qerror!("Duplicate push stream.");
+                    Err(Error::HttpId)
+                }
+            },
+        }
+    }
+
+    fn check_push_id_and_create_stream_state(&mut self, push_id: u64) -> Res<()> {
+        // Check if push id is greater than what we allow.
+        if push_id > self.current_max_push_id {
+            qerror!("Push id is greater than current_max_push_id.");
+            return Err(Error::HttpId);
+        }
+
+        while self.next_push_id_to_open <= push_id {
+            self.push_streams
+                .insert(self.next_push_id_to_open, PushState::Init);
+            self.next_push_id_to_open += 1;
+        }
```

Comment:
You can use `VecDeque::resize` for this (but I would move the function to the collection wrapper.